src/calibre/ebooks/metadata/odt.py

```diff
@@ -220,6 +224,28 @@ def get_metadata(stream, extract_cover=True):
 
     return mi
 
+def get_meta_doc_props(mi):
+    metaFields = {}
+    metaFields[fields.get('title')] = mi.title
+    metaFields[fields.get('creator')] = authors_to_string(mi.authors)
+    if mi.tags:
+        metaFields[fields.get('keyword')] = ', '.join(mi.tags)
+    if mi.comments:
+        metaFields[fields.get('description')] = mi.comments
+    if mi.languages:
+        l = canonicalize_lang(mi.languages[0])
+        metaFields[fields.get('language')] = lang_as_iso639_1(l) or l
+    return metaFields
+
+def set_metadata(stream, mi):
+    from calibre.utils.zipfile import safe_replace
+    metaFields = get_meta_doc_props(mi)
+
+    zin, odfs = get_odf_meta_parsed(stream, addfields=metaFields, deletefields=metaFields)
+    stream.seek(os.SEEK_SET)
+    safe_replace(stream, "meta.xml", io.BytesIO(str(odfs.meta())))
```

Comment:
As part of the effort to port to python3, we always use unicode_type from polyglot.builtins -- never str(). In this case it reveals a bug -- you cannot use a python3 str() when writing to a BytesIO buffer.

Instead, since meta() returns a MetaCollector.content() (which is a unicode_literals string value), you should be using .encode('utf-8'). Using str() to cast to bytes only on python2 is wrong.

Reply:
Ok, thanks, I fix-it!